pallets/dynamic-evm-base-fee/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+                };
+
+                // It's possible current base fee per gas is outside of the allowed range.
+                // This can & will happen when this solution is deployed on live networks.
+                //
+                // In such scenario, we will discard the lower & upper bounds configured in the runtime.
+                // Once these bounds are reached ONCE, the runtime logic will prevent them from going out of bounds again.
+                let apply_configured_bounds = old_bfpg >= T::MinBaseFeePerGas::get()
+                    && old_bfpg <= T::MaxBaseFeePerGas::get();
+                let (lower_limit, upper_limit) = if apply_configured_bounds {
+                    (
+                        T::MinBaseFeePerGas::get().max(old_bfpg.saturating_sub(max_step)),
+                        T::MaxBaseFeePerGas::get().min(old_bfpg.saturating_add(max_step)),
+                    )
+                } else {
+                    (
+                        old_bfpg.saturating_sub(max_step),
+                        old_bfpg.saturating_add(max_step),
+                    )
+                };
+
+                // Calculate ideal new 'base_fee_per_gas' according to the formula
+                let ideal_new_bfpg = T::AdjustmentFactor::get()
+                    // Weight factor should be multiplied first since it's a larger number, to avoid precision loss.
+                    .saturating_mul_int(T::WeightFactor::get())
+                    .saturating_mul(25)
+                    .saturating_div(98974);
+
+                // Clamp the ideal value in between the allowed limits
+                *base_fee_per_gas = U256::from(ideal_new_bfpg).clamp(lower_limit, upper_limit);
+            })
+        }
+
+        fn integrity_test() {
+            assert!(T::MinBaseFeePerGas::get() <= T::MaxBaseFeePerGas::get(),
+                "Minimum base fee per gas has to be equal or lower than maximum allowed base fee per gas.");
+
+            assert!(T::DefaultBaseFeePerGas::get() >= T::MinBaseFeePerGas::get(),
+                "Default base fee per gas has to be equal or higher than minimum allowed base fee per gas.");
+            assert!(T::DefaultBaseFeePerGas::get() <= T::MaxBaseFeePerGas::get(),
+                "Default base fee per gas has to be equal or lower than maximum allowed base fee per gas.");
+
+            assert!(T::MaxBaseFeePerGas::get() <= U256::from(u128::MAX),
+                "Maximum base fee per gas has to be equal or lower than u128::MAX, otherwise precision loss will occur.");
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// `root-only` extrinsic to set the `base_fee_per_gas` value manually.
+        /// The specified value has to respect min & max limits configured in the runtime.
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::set_base_fee_per_gas())]
+        pub fn set_base_fee_per_gas(origin: OriginFor<T>, fee: U256) -> DispatchResult {
```

Comment:
It's useful for tests.
Especially for RPC tests, where we can test behavior under different values.

For live networks, it could be useful to set it to some fixed value if algorithm goes wrong.

But in general, it's not supposed to be used in production.
Hence the `root-only` approach.